src/Microsoft.DotNet.Darc/Darc/Program.cs

```diff
@@ -67,6 +102,138 @@ private static int RunOperation(CommandLineOptions opts)
         }
     }
 
+    private static void RegisterServices(IServiceCollection services, CommandLineOptions options)
+    {
+        // Because the internal logging in DarcLib tends to be chatty and non-useful,
+        // we remap the --verbose switch onto 'info', --debug onto highest level, and the
+        // default level onto warning
+        LogLevel level = LogLevel.Warning;
+        if (options.Debug)
+        {
+            level = LogLevel.Debug;
+        }
+        else if (options.Verbose)
+        {
+            level = LogLevel.Information;
+        }
+
+        services ??= new ServiceCollection();
+        services.AddLogging(b => b
+            .AddConsole(o => o.FormatterName = CompactConsoleLoggerFormatter.FormatterName)
+            .AddConsoleFormatter<CompactConsoleLoggerFormatter, SimpleConsoleFormatterOptions>()
+            .SetMinimumLevel(level));
+
+        services.AddSingleton(options);
+        services.TryAddSingleton<IFileSystem, FileSystem>();
+        services.TryAddSingleton<IRemoteFactory, RemoteFactory>();
+        services.TryAddTransient<IProcessManager>(sp => ActivatorUtilities.CreateInstance<ProcessManager>(sp, options.GitLocation));
+        services.TryAddSingleton(sp => RemoteFactory.GetBarClient(options, sp.GetRequiredService<ILogger<BarApiClient>>()));
+        services.TryAddSingleton<IBasicBarClient>(sp => sp.GetRequiredService<IBarApiClient>());
+        services.TryAddTransient<ILogger>(sp => sp.GetRequiredService<ILogger<Operation>>());
+        services.TryAddTransient<ITelemetryRecorder, NoTelemetryRecorder>();
+        services.TryAddTransient<IGitRepoFactory>(sp => ActivatorUtilities.CreateInstance<GitRepoFactory>(sp, Path.GetTempPath()));
+        services.Configure<AzureDevOpsTokenProviderOptions>(o =>
+        {
+            o["default"] = new AzureDevOpsCredentialResolverOptions
+            {
+                Token = options.AzureDevOpsPat,
+                FederatedToken = options.FederatedToken,
+                DisableInteractiveAuth = options.IsCi,
+            };
+        });
+        services.TryAddSingleton<IAzureDevOpsTokenProvider, AzureDevOpsTokenProvider>();
+        services.TryAddSingleton(s =>
+            new AzureDevOpsClient(
+                s.GetRequiredService<IAzureDevOpsTokenProvider>(),
+                s.GetRequiredService<IProcessManager>(),
+                s.GetRequiredService<ILogger>())
+        );
+        services.TryAddSingleton<IAzureDevOpsClient>(s =>
+            s.GetRequiredService<AzureDevOpsClient>()
+        );
+        services.TryAddSingleton<IRemoteTokenProvider>(_ => new RemoteTokenProvider(options.AzureDevOpsPat, options.GitHubPat));
+        services.TryAddSingleton<CommandLineOptions>(_ => options);
+    }
+
+    private static void RegisterVmrServices(ServiceCollection services, VmrCommandLineOptions vmrOptions)
```

Comment:
I must say, I don't like this.

It made more sense if this was tied to the inheritance model and was in the VMR options (or potentially operation).

Reply:
Maybe a better model would be to have a `RegisterServices` method in the `CommandLineOptions` class, which would register all non vmr classes, and then override it in VMR options, which would register all vmr stuff, and then call the base class method

Reply:
The way it was before, you could have already registered additional stuff in the options I think. Or maybe the operation could pass the service collection to base(). One of them could do it for sure